libstdc++-v3: enable std::thread for zephyr via posix

cfriedt opened the pull request:
Add scaffolding to set required zephyr-specific bits for gcc in order to enable standard C++ threads, mutexes, condition variables, etc, using the POSIX API as a foundation.

Note: it is not possible to run compile and / or link tests under libstdc++-v3 at this time.

Commit:
libstdc++-v3: support for std::thread mutex etc via gthr-posix.h

Use the existing gthr-posix.h configuration for gthreads to
enable support in Zephyr for C++ std::thread, std::mutex,
std::condition_variable, std::binary_semaphore,
std::counting_semaphore, and more.

Signed-off-by: Chris Friedt <cfriedt@meta.com>

cfriedt commented:
Tested to work with zephyrproject-rtos/zephyr#43729 as of commit `cb7b650`

<img width="882" alt="Screenshot 2023-12-01 at 4 44 38 PM" src="https://github.com/zephyrproject-rtos/gcc/assets/983494/902837bc-9a7a-4d01-afb1-62b668e37c30">

The linked PR includes support and tests covering `std::thread`, `std::mutex`, `std::condition_variable`, and quite a bit more.

stephanosio commented on libstdc++-v3/configure.ac:
```diff
@@ -212,10 +212,16 @@ GLIBCXX_CHECK_C99_TR1
 
 # For gettimeofday support.
-GLIBCXX_CHECK_GETTIMEOFDAY
+AC_DEFINE(_GLIBCXX_USE_GETTIMEOFDAY)
```
This is too brute force IMO. `GLIBCXX_CHECK_GETTIMEOFDAY` should define `_GLIBCXX_USE_GETTIMEOFDAY` anyway if it is available.

cfriedt replied:
Fixed

stephanosio commented on libstdc++-v3/configure.ac:
```diff
@@ -212,10 +212,16 @@ GLIBCXX_CHECK_C99_TR1
 
 # For clock_gettime, nanosleep and sched_yield support.
-GLIBCXX_ENABLE_LIBSTDCXX_TIME
+AC_DEFINE(_GLIBCXX_USE_CLOCK_GETTIME_SYSCALL)
+AC_DEFINE(_GLIBCXX_USE_CLOCK_MONOTONIC)
+AC_DEFINE(_GLIBCXX_USE_CLOCK_REALTIME)
+AC_DEFINE(_GLIBCXX_USE_SCHED_YIELD)
+AC_DEFINE(_GLIBCXX_USE_NANOSLEEP)
+AC_DEFINE(HAVE_SLEEP)
+AC_DEFINE(HAVE_USLEEP)
```
We should not be brute-forcing global configs like this.

[`GLIBCXX_ENABLE_LIBSTDCXX_TIME`](https://github.com/zephyrproject-rtos/gcc/blob/zephyr-gcc-12.2.0/libstdc%2B%2B-v3/acinclude.m4#L1332) itself should be modified to set Zephyr-specific configs that apply when the target is Zephyr.

stephanosio commented on libstdc++-v3/configure.ac:
```diff
@@ -481,8 +487,11 @@ GLIBCXX_CHECK_X86_RDSEED
 # This depends on GLIBCXX_ENABLE_SYMVERS and GLIBCXX_IS_NATIVE.
 GLIBCXX_CONFIGURE_TESTSUITE
 
-# For gthread support.  Depends on GLIBCXX_ENABLE_SYMVERS.
-GLIBCXX_CHECK_GTHREADS
+# For gthread support.
+AC_DEFINE(_GTHREAD_USE_MUTEX_TIMEDLOCK)
+AC_DEFINE(_GLIBCXX_HAS_GTHREADS)
+AC_DEFINE(_GLIBCXX_USE_PTHREAD_RWLOCK_T)
+AC_DEFINE(_GLIBCXX_HAVE_POSIX_SEMAPHORE)
```
same here

stephanosio reviewed (changes requested):
In addition to the comments below, need to regenerate configure script as well.

cfriedt commented:
> In addition to the comments below, need to regenerate configure script as well.

@stephanosio - I just saw that some zephyr things were hard-coded in the `./configure` script (which should really be generated and not edited).

Are there some Zephyr-SDK specific steps to regenerating it? 

E.g. what about editing `gnu-config` (aka `config`) items so that zephyr is recognized as an OS. That is (mainly) what should eventually be upstreamed.

Otherwise, it would be like
* regenerate `./configure`
* copy out all new sections relevant to the change
* revert to the previous `./configure`
* manually edit / update new sections relevant to the change

I can imagine that is.. quite a bit of work.. and likely makes life significantly harder than it needs to be

Flameeyes commented:
My recommendation would be to patch autoconf so that it knows about Zephyr, and then use this patched autoconf to regenerate the `configure` script, even if the autoconf patch is not quite ready for upstream.

Besides giving a head start to upstreaming, this would mean never having to touch the generated files. Which is a lot more resilient.

stephanosio commented:
> @stephanosio - I just saw that some zephyr things were hard-coded in the `./configure` script (which should really be generated and not edited).
>
> Are there some Zephyr-SDK specific steps to regenerating it?

It is not hard-coded. `configure` script is generated using autoreconf from `configure.ac` as done in the upstream GCC (see https://gcc.gnu.org/wiki/Regenerating_GCC_Configuration).

> E.g. what about editing `gnu-config` (aka `config`) items so that zephyr is recognized as an OS. That is (mainly) what should eventually be upstreamed.

I have already submitted an upstream patch and it was accepted a while ago (it should already be part of the latest GCC `config.sub`) -- see https://git.savannah.gnu.org/cgit/config.git/commit/?id=401e1a7dd1e082fd110ba9df169d2de9ec2bab74.

Of course, that only makes it "recognise" the target, the rest of the target support still need to be added to GCC and Binutils (which is basically what https://github.com/zephyrproject-rtos/sdk-ng/issues/350 describes).

stephanosio commented on libstdc++-v3/acinclude.m4:
```diff
@@ -1394,6 +1394,14 @@ AC_DEFUN([GLIBCXX_ENABLE_LIBSTDCXX_TIME], [
       uclinux*)
         ac_has_nanosleep=yes
         ac_has_sched_yield=yes
+      zephyr*)
```
This will not work since this switch is for `target_os` and `zephyr` is currently not an "OS" (it is a "vendor"). Will need to check for `${target}` here instead.

cfriedt replied:
Is that upstream too?

stephanosio commented on libstdc++-v3/acinclude.m4:
```diff
@@ -1589,12 +1597,21 @@ AC_DEFUN([GLIBCXX_CHECK_GETTIMEOFDAY], [
 
   AC_MSG_CHECKING([for gettimeofday])
 
+  case "${target_os}" in
+    zephyr*)
+      ac_has_gettimeofday=yes
+      ;;
```
same deal

stephanosio commented on libstdc++-v3/acinclude.m4:
```diff
@@ -3964,6 +3982,17 @@ dnl
 AC_DEFUN([GLIBCXX_CHECK_GTHREADS], [
   GLIBCXX_ENABLE(libstdcxx-threads,auto,,[enable C++11 threads support])
 
+  case "${target_os}" in
```
ditto

stephanosio reviewed (commented):
Still need to regenerate `configure`.

cfriedt commented:
> I have already submitted an upstream patch and it was accepted a while ago (it should already be part of the latest GCC `config.sub`) -- see https://git.savannah.gnu.org/cgit/config.git/commit/?id=401e1a7dd1e082fd110ba9df169d2de9ec2bab74.

Oh nice - I guess my autoconf needs to be updated (has autoconf had a full release since then?).

> Of course, that only makes it "recognise" the target, the rest of the target support still need to be added to GCC and Binutils (which is basically what [zephyrproject-rtos/sdk-ng#350](https://github.com/zephyrproject-rtos/sdk-ng/issues/350) describes).

Awesome - I almost forgot about that issue. It's been a while. Making progress though.

Commit:
libstdc++-v3: update configure for std::thread via gthr-posix.h

Regenerate the ./configure script with zephyr-specific conditions
to enable std::thread, std::mutex, std::condition_variable,
std::binary_semaphore, std::counting_semaphore, etc.

Signed-off-by: Chris Friedt <cfriedt@meta.com>

cfriedt commented:
> In addition to the comments below, need to regenerate configure script as well.

I know that there are probably reasons for doing it this way, but it's not at all sustainable to check in / modify generated code. Created https://github.com/zephyrproject-rtos/sdk-ng/issues/723 to track the issue.

Commit:
include zephyr bits to support cross-compilation

This change adds a zephyr-bits.h header file which includes the
basic definitions required to support ISO C++ thread, mutex,
condition variables, and more via POSIX. For additional details
see libgcc/gthr-posix.h .

Note: this header should not be exported; it is solely for the
purpose of enabling C++ threads in the Zephyr SDK when building
against newlib or picolibc.

Signed-off-by: Chris Friedt <cfriedt@meta.com>

Commit:
update configure to include zephyr-bits.h

Touch-up the generated configure script to reflect changes in
configure.ac .

Signed-off-by: Chris Friedt <cfriedt@meta.com>

cfriedt commented:
@stephanosio 

Building the Zephyr SDK manually is successful with the current changes in zephyrproject-rtos/gcc#25 and zephyrproject-rtos/sdk-ng#722 but unfortunately CI fails without really having any kind of decent diagnostics or debug information.

[build.log.xz.txt](https://github.com/zephyrproject-rtos/zephyr/files/13591305/build.log.xz.txt) (`xz -d < build.log.xz.txt > build.log.txt`)

Not really sure where to go from here.

cfriedt commented:
Closing this in favour of the c11 altrnative
